### PR TITLE
audit: Port line_problems to rubocop and add tests part 4

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----
#569 

This almost completely ports `line_problems` . This is a huge change, and I wrote tests for every rule that has been ported. Feel free to suggest additional tests and I would add them. 
Opening a PR so it can be reviewed as there are no conflicts now.

### CODE REVIEW

File: Library/.rubocop.yml [hunk 1]
```diff
@@ -118,6 +118,10 @@ Style/Documentation:
 Style/Encoding:
   Enabled: true
 
+# use spaces for indentation; detect tabs
+Style/Tab:
+  Enabled: true
```
Comment (reitermarkus) [inline Library/.rubocop.yml]: Is this not enabled by default?

File: Library/Homebrew/test/rubocops/lines_cop_spec.rb [hunk 2]
```diff
@@ -419,7 +419,7 @@ class Foo < Formula
       end
     end
 
-    it "with ENV.universal_binary" do
+    it "with ENV.universal_binary 2" do
```
Comment (reitermarkus) [inline Library/Homebrew/test/rubocops/lines_cop_spec.rb]: Why `2`? Also please create full sentences when using `it`, or use `specify` instead of `it`.

**Comment by DomT4** [comment]
Small note: This PR semi-bumps into https://github.com/Homebrew/brew/pull/3293. I can always tweak that PR afterwards if this goes in first though, not a huge issue.

**Comment by MikeMcQuaid** [comment]
@GauthamGoli Needs rebased now https://github.com/Homebrew/brew/pull/3293 was merged, sorry. Also, some `brew style` failures here. Let us know when comments/style/tests are addressed and I'll try and do some tests and get this merged soon after.

File: Library/Homebrew/rubocops/lines_cop.rb [hunk 3]
```diff
@@ -127,8 +238,124 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
...
+
+          find_instance_method_call(body_node, "Dir", :[]) do |method|
+            path = parameters(method).first
+            next if !path.str_type?
+            next unless match = regex_match_group(path, /^[^\*{},]+$/)
+            problem "Dir([\"#{string_content(path)}\"]) is unnecessary; just use \"#{match[0]}\""
+          end
+
+
+          fileUtils_methods= Regexp.new(FileUtils.singleton_methods(false).map { |m| '(?-mix:^'+Regexp.escape(m)+'$)' }.join '|')
```
Comment (DomT4) [inline Library/Homebrew/rubocops/lines_cop.rb]: I know it's fussy, but for consistency there should probably be a space between the `_methods` and the `=` here?

Comment (GauthamGoli) [reply Library/Homebrew/rubocops/lines_cop.rb]: Its fine. Thanks! 👍

Comment (reitermarkus) [reply Library/Homebrew/rubocops/lines_cop.rb]: And don‘t use camelCase in Ruby.

File: Library/Homebrew/test/rubocops/lines_cop_spec.rb (hunk 2 above)
Comment (GauthamGoli) [reply Library/Homebrew/test/rubocops/lines_cop_spec.rb]: Tried replacing with somewhat better phrases.

**Comment by MikeMcQuaid** [comment]
Nice work here @GauthamGoli, thanks again!

**Comment by ilovezfs** [comment]
This is causing
```
Error: 1 problem in 1 formula
infer:
  * C: 54: col 85: Use "#{ENV.cc}" instead of hard-coding "clang"
```
on https://github.com/Homebrew/homebrew-core/pull/19646

See https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/11525/version=high_sierra/testReport/junit/brew-test-bot/high_sierra/audit_infer___online/

**Pull request merged**